### PR TITLE
[CREATE Part A (updated)] Modification to Copy Circuit

### PR DESCRIPTION
**Pull request by ed255**
### Description

NOTE: This is an updated version of https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1356

This is Part A of a 3 part pull request to add support for `CREATE`/`CREATE2` opcodes.

Part A: https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1356
Part B: https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1357
Part C: https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1358

### Issue Link

#1130 

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents
 
As part of the bigger additions needed for the `CREATE`/`CREATE2` opcodes' gadget, this PR adds support for the copy circuit to "always" have a value accumulator field `value_acc`.

### Rationale

We need a value accumulator (of the random linear combination) in order to get the `RLC(bytes)` for the bytes copied from `Memory` to `Bytecode` (specifically the init code). This RLC is later used to do a lookup to the Keccak table to check the value of `keccak256(init_code)`.

### How Has This Been Tested?

The existing tests for copy circuit pass for the updated constraints on the copy circuit.

### CODE REVIEW

**Comment by ed255** [comment]
This PR is a continuation of https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1356
The requests and comments have not been addressed yet, that's the next step before merging.

File: zkevm-circuits/src/copy_circuit.rs
Comment (han0110) [inline zkevm-circuits/src/copy_circuit.rs:327]: We can remove line 326 since `q_enable = 1` when `q_step = 1`.

File: zkevm-circuits/src/copy_circuit.rs
Comment (han0110) [inline zkevm-circuits/src/copy_circuit.rs:297]: This'd be a little bit cheaper
```suggestion
                not::expr(
                    and::expr([
                        meta.query_advice(is_last, Rotation::next()),
                        meta.query_advice(is_pad, Rotation::cur()),
                    ]),
                ),
```

File: zkevm-circuits/src/copy_circuit.rs
Comment (han0110) [inline zkevm-circuits/src/copy_circuit.rs:304]: It seems we don't have constraint to enforce for the first step of each copy event, the `value_acc_rlc` should be equal to `value`. A malicious prover now can start with any value.

File: zkevm-circuits/src/copy_circuit.rs
Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs:327]: Sorry, I didn't see any place constraint `q_enable = 1` when `q_step = 1`. Could you please show me where?

File: zkevm-circuits/src/copy_circuit.rs
Comment (roynalnaruto) [reply zkevm-circuits/src/copy_circuit.rs:304]: You're right. After changing the layout I must have forgotten to re-add that constraint. Thank you for noticing!

File: zkevm-circuits/src/copy_circuit.rs
Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs]: fixed in [d2bdc68](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1419/commits/d2bdc68414d218732a56d4e1d038b1c11859c9e6)

File: zkevm-circuits/src/copy_circuit.rs
Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs:311]: fixed in [d2bdc68](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1419/commits/d2bdc68414d218732a56d4e1d038b1c11859c9e6)

File: zkevm-circuits/src/copy_circuit.rs
Comment (han0110) [reply zkevm-circuits/src/copy_circuit.rs:334]: Since both `q_enable` and `q_step` are fixed column, and from the design (or assignment) `q_enable` is for every row (except blinding factors) and `q_step` is for every 2-row (except blinding factors).

Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs]: fixed in [d591f47](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1419/commits/d591f47ab9558031eacf11e4e04b148685db7654)

File: zkevm-circuits/src/copy_circuit.rs
Comment (han0110) [reply zkevm-circuits/src/copy_circuit.rs:311]: The following question is, is there any constraint to enforce `is_first` to be `1` when we are entering next copy event? Since I couldn't find an explicit one for that, and after I change the `is_first` assignment to all `0` the tests are still passing.

File: zkevm-circuits/src/copy_circuit.rs
Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs]: @han0110 , it should be `or` instead of `and`. Could you help to confirm? fixed in [b696d4a](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1419/commits/b696d4a9bf92904ad71c8234fd21a9986091eb40)
```rust
                not::expr(
                    or::expr([
                        meta.query_advice(is_last, Rotation::next()),
                        meta.query_advice(is_pad, Rotation::cur()),
                    ]),
                ),
```

File: zkevm-circuits/src/copy_circuit.rs
Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs]: >  from the design (or assignment) q_enable is for every row (except blinding factors) and q_step is for every 2-row (except blinding factors).

I could understand from the design we'll assign `q_enable` for every row but I thought we can't trust assignment part. Which means malicious provers could set `q_enable = 0` when `q_step = 1` and the constraints here will be bypassed.

Comment (han0110) [reply zkevm-circuits/src/copy_circuit.rs]: But for `q_enable` and `q_step` they are both preprocess polynomial, so I think we can trust the assignment, right?

Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs]: yup, make sense, you're right! thanks.

File: zkevm-circuits/src/copy_circuit.rs
Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs:311]: https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1432 created for `is_first` and `is_last`

File: zkevm-circuits/src/copy_circuit.rs
Comment (ed255) [inline zkevm-circuits/src/copy_circuit.rs:255]: The expression from here may be causing the final gate degree ti increase more than 9 (and that's why tests are not passing).  A simple solution would be to split this into 2 constraints. Following this pattern:
`if A or B then X` -> `if A then X`, `if B then X`.

Another thing that comes to my mind is that `or::expr` returns a boolean, but to build a selector expression just having `0` when false and `!= 0` when true is enough, so perhaps you could replace the `or` by a `+`.  This would give `2` when both expressions are true, but that's ok.

**Review by ed255 — COMMENTED** [review]
Currently the tests are not passing due to a degree increase of the gates expressions, take a look at my comment for a possible fix.  Following Han observation about the missing constraints on `is_first` and `is_last` I think a careful review of this circuit could be really useful!  I suggest that we resolve the underconstraints in a future PR (Kimi already opened an issue for that), and move on with this PR once the tests pass to unblock the CREATE.

File: zkevm-circuits/src/copy_circuit.rs
Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs:54]: This was @roynalnaruto's fix and not sure the reason behind. But it seems solved the issue you mentioned.

Comment (han0110) [reply zkevm-circuits/src/copy_circuit.rs:54]: I think to have an extra `value_acc_rlc` is to support `memory <-> bytecode` copy and rlc at the same time without changing other parts too much.

And it seems to also resolve #971, so we could try to move the phase of `value` from second to first and see if it works as expected.

File: zkevm-circuits/src/copy_circuit.rs
Comment (KimiWu123) [reply zkevm-circuits/src/copy_circuit.rs]: Nice idea! It looks nice and clean now, fixed in [52d480a](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1419/commits/52d480a1dcade244606b6d2598abd1f756a80c1d)

**Review by han0110 — APPROVED** [review]
LGTM now! Thanks for addressing all the comments.

**Review by KimiWu123 — APPROVED** [review]

**Pull request merged**
